### DISCUSSION TITLE
Markdown render match errors

### DISCUSSION THREAD

**Discussion by celinehaddad** [remarkjs/remark#1130] [Q&A] [answered]
Hey everyone, 

I am trying to render a custom CodeBlock component inside of a custom collapsed Note component in markdown. The regex for collapsed note is `regex = /\?\?\?\s*(?<title>[^(\n)]*)(\n|\n\n)(?<message>.[^\t]*)/;` & it matches these markdown input formats:

![Screen Shot 2023-03-13 at 9 55 53 PM](https://user-images.githubusercontent.com/68402310/224874160-41b18eb0-ec82-4a4f-87c9-439524b68a31.png)

Which render like this: 
![Screen Shot 2023-03-13 at 9 55 41 PM](https://user-images.githubusercontent.com/68402310/224874158-7cea4a24-5094-4cdd-a760-2f3bf536086a.png)

1. I need to support the format in lines 11-15, but only lines 18-21 render as designed. The regex `(\n|\n\n)` between title and message groups should also match the markdown format in lines 11-15, so this is not a regex issue. These are the remark and rehype plugins I am using: 

![Screen Shot 2023-03-13 at 10 17 21 PM](https://user-images.githubusercontent.com/68402310/224875649-e608b5a7-cb52-4d1c-91f9-e8a21f234e1c.png)

Do you see any issues with my plugins that could prevent lines 11-15 from rendering correctly? 

2. I need to match the markdown format for CodeBlock inside collapsed Note in lines 48-55: 
![Screen Shot 2023-03-13 at 10 39 19 PM](https://user-images.githubusercontent.com/68402310/224878713-0245416a-d4e3-4fbf-a4f1-1a95d9abf921.png)
Which renders like this: 
![Screen Shot 2023-03-13 at 10 40 47 PM](https://user-images.githubusercontent.com/68402310/224879098-c888985a-ba98-4dea-9c2e-0d3f2159747c.png)

We need the tabs on the CodeBlock for this because without the tabs, the CodeBlock renders outside of the collapsed note. 
![Screen Shot 2023-03-13 at 10 37 53 PM](https://user-images.githubusercontent.com/68402310/224879432-30768d46-adf9-4c8e-a869-3b1381fe784b.png)

Best practices in mind, how would you support something like this?

**Comment by Murderlon** [comment, marked as answer]
> Best practices in mind, how would you support something like this?

I think this is the most important question. [Extending markdown](https://github.com/micromark/micromark/#extending-markdown) is almost always the least favorable option. But if you do, you should do it at the parser level and not with regex. One of the main advantages of using the unified ecosystem is that you can work with AST's, which is far less error prone than regex, especially with all the utilities and plugins we provide. 

Taking a step back, it might be good to figure that you are trying to achieve with custom syntax. Most likely, you can solve it with [`remark-directive`](https://github.com/remarkjs/remark-directive) and [`remark-code-extra`](https://github.com/s0/remark-code-extra).

**Comment by celinehaddad** [comment]
Thank you for getting back to me @Murderlon. Got it- AST's only, no regex. I've explored and integrated `remark-directive`. However, I seem to have an issue, potentially with the handler in `remarkRehypeOptions`. I scrapped the custom plugin that we had with regex, and added this plugin:

![Screen Shot 2023-03-17 at 2 02 50 AM](https://user-images.githubusercontent.com/68402310/225826154-2a524567-4d58-46f8-8207-b87b9d1e8147.png)

Which is successfully applying the class & title attributes to the html when I inspect the browser elements (I'm aware I haven't yet added support for a Note with a CodeBlock inside for the third example): 

![Screen Shot 2023-03-17 at 2 24 03 AM](https://user-images.githubusercontent.com/68402310/225828707-95604a3d-845c-4367-960c-72cdb2ab44b4.png)

Here's the new markdown format for that html^ which isn't rendering as my Note component: 

![Screen Shot 2023-03-17 at 2 00 47 AM](https://user-images.githubusercontent.com/68402310/225826262-492669d0-59b8-4c7d-ae80-f2050f5b0319.png)

Here is my Note component: 

![Screen Shot 2023-03-17 at 2 13 26 AM](https://user-images.githubusercontent.com/68402310/225827083-35c276a4-4c84-4ecc-b4fa-c62a825be763.png)

I updated the remark rehype plugins list- the handler on line 37 is where I suspect the issue to be: 
![Screen Shot 2023-03-17 at 2 03 27 AM](https://user-images.githubusercontent.com/68402310/225827634-3f2c3fb0-6fbb-4810-8f20-52b9b9c9dde8.png)

Do you see anything I could be missing?

**Reply by wooorm** [reply]
P.S. it’s easier if you paste code as markdown instead of images btw, more easy to copy/paste, more accessible too.

Note that attributes in the directives don’t need commas! Similar to how you don’t put commas in HTML tags between attributes either. Now those commas you use are actually present in the class names.

> which isn't rendering as my Note component:

> Do you see anything I could be missing?

I see that you define a `Note` component. I don’t see any code that *uses* that `Note` variable.
So I don’t see how some variable that you never use should magically be used?

**Reply by Murderlon** [reply]
Something I forgot to say since you mentioned collapsable notes, is that you could also use HTML for this use case: `<details>`. This is what [docusaurus does](https://docusaurus.io/docs/markdown-features#details) for instance (with [this component](https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-common/src/components/Details/index.tsx#L45)). 

I think the approach depends a bit on who authors the markdown, non tech savvy people likely don't want to write HTML?

**Reply by celinehaddad** [reply]
Hey @wooorm, thanks will do. That explains why my test classes stopped working after adding the title attribute in the markdown, I removed the commas. Meant to share this as well- I use the note variable here: 

```typescript
// ./components/index.ts:
import CodeBlock from "./CodeBlock";
import TabList from "./TabList";
import Note from "./Note";

export default {
  code: CodeBlock,
  tablist: TabList,
  note: Note,
};
```

@Murderlon thanks for sharing those links. After taking a look, I want to keep it so that the tech writers that add to these docs will only have to write markdown, and super minimal HTML like br elements.

**Reply by celinehaddad** [reply]
Also, added `node.type = "note";` on line 23 of our remarkNote plugin: 

```typescript
import { visit } from "unist-util-visit";
import { h } from "hastscript";

function remarkNote() {
  return (tree: any) => {
    visit(tree, (node: any) => {
      if (
        node.type === "textDirective" ||
        node.type === "leafDirective" ||
        node.type === "containerDirective"
      ) {
        if (node.name !== "note") return;

        const data = node.data || (node.data = {});
        const tagName = node.type === "textDirective" ? "span" : "div";

        data.hName = tagName;
        data.hProperties = h(tagName, node.attributes).properties;
        node.title = node.attributes.title;
        node.type = "note";
      }
    });
  };
}

export default remarkNote;
```

**Reply by wooorm** [reply]
`node.type` doesn’t really do anything.

But `data.hName` *is* the type that will be used. That‘s an element tag name. Which you set to `span` or `div`.
*That’s* the key that will be looked up in `components`.
Say you set that to `p`, then you can have a `components.p` which will be used.

You can set that to `note`, and then use your `components.note` component.
As it’s supposed to be an HTML tag name, and `note` isn’t a correct HTML tag name, you could also use a custom element name (tag name with `-`), such as `my-note`, and then have a component called `my-note`

**Reply by celinehaddad** [reply]
@wooorm got it, thank you so much! this worked for me.
